### PR TITLE
Added test to validate resource limits for guest

### PR DESCRIPTION
**Pull request by sathnaga**
Added test to validate resource limits(i.e cpu, memory,
network, disk etc) for guest
Signed-off-by: Satheesh Rajendran <sathnaga@linux.vnet.ibm.com>

### CODE REVIEW

**Comment by sathnaga** [comment]
still testing out, not yet ready for review, will inform once done
 ``` (1/6) guest_cpu.limittest.qemu.qcow2.virtio_scsi.smp2.virtio_net.RHEL.7.devel.ppc64le.powerkvm-libvirt.limit.with_graphics.max_cpu: PASS (73.06 s)
 (2/6) guest_cpu.limittest.qemu.qcow2.virtio_scsi.smp2.virtio_net.RHEL.7.devel.ppc64le.powerkvm-libvirt.limit.with_graphics.max_mem: PASS (121.82 s)
 (3/6) guest_cpu.limittest.qemu.qcow2.virtio_scsi.smp2.virtio_net.RHEL.7.devel.ppc64le.powerkvm-libvirt.limit.with_graphics.max_mem_cpu: PASS (140.45 s)
 (4/6) guest_cpu.limittest.qemu.qcow2.virtio_scsi.smp2.virtio_net.RHEL.7.devel.ppc64le.powerkvm-libvirt.limit.without_graphics.max_cpu: PASS (58.93 s)
 (5/6) guest_cpu.limittest.qemu.qcow2.virtio_scsi.smp2.virtio_net.RHEL.7.devel.ppc64le.powerkvm-libvirt.limit.without_graphics.max_mem: PASS (131.91 s)
 (6/6) guest_cpu.limittest.qemu.qcow2.virtio_scsi.smp2.virtio_net.RHEL.7.devel.ppc64le.powerkvm-libvirt.limit.without_graphics.max_mem_cpu: PASS (144.78 s)
RESULTS    : PASS 6 | ERROR 0 | FAIL 0 | SKIP 0 | WARN 0 | INTERRUPT 0 | CANCEL 0
JOB TIME   : 677.55 s```

**Comment by sathnaga** [comment]
```
 (01/18) guest_import.qemu.qcow2.virtio_scsi.smp2.virtio_net.RHEL.7.devel.ppc64le.powerkvm-qemu.unattended_install.import.import.default_install.aio_native: PASS (35.20 s)
 (02/18) guest_limit.qemu.qcow2.virtio_scsi.smp2.virtio_net.RHEL.7.devel.ppc64le.powerkvm-libvirt.limit.with_graphics.max_cpu: PASS (63.52 s)
 (03/18) guest_limit.qemu.qcow2.virtio_scsi.smp2.virtio_net.RHEL.7.devel.ppc64le.powerkvm-libvirt.limit.with_graphics.max_mem: PASS (147.69 s)
 (04/18) guest_limit.qemu.qcow2.virtio_scsi.smp2.virtio_net.RHEL.7.devel.ppc64le.powerkvm-libvirt.limit.with_graphics.max_mem_cpu: PASS (203.26 s)
 (05/18) guest_limit.qemu.qcow2.virtio_scsi.smp2.virtio_net.RHEL.7.devel.ppc64le.powerkvm-libvirt.limit.with_graphics.max_disk: PASS (64.33 s)
 (06/18) guest_limit.qemu.qcow2.virtio_scsi.smp2.virtio_net.RHEL.7.devel.ppc64le.powerkvm-libvirt.limit.with_graphics.max_network: PASS (61.03 s)
 (07/18) guest_limit.qemu.qcow2.virtio_scsi.smp2.virtio_net.RHEL.7.devel.ppc64le.powerkvm-libvirt.limit.with_graphics.max_disk_network: PASS (77.41 s)
 (08/18) guest_limit.qemu.qcow2.virtio_scsi.smp2.virtio_net.RHEL.7.devel.ppc64le.powerkvm-libvirt.limit.with_graphics.max_cpu_disk_network: PASS (93.12 s)
 (09/18) guest_limit.qemu.qcow2.virtio_scsi.smp2.virtio_net.RHEL.7.devel.ppc64le.powerkvm-libvirt.limit.with_graphics.max_cpu_mem_disk_network: PASS (226.23 s)
 (10/18) guest_limit.qemu.qcow2.virtio_scsi.smp2.virtio_net.RHEL.7.devel.ppc64le.powerkvm-libvirt.limit.without_graphics.max_cpu: PASS (71.17 s)
 (11/18) guest_limit.qemu.qcow2.virtio_scsi.smp2.virtio_net.RHEL.7.devel.ppc64le.powerkvm-libvirt.limit.without_graphics.max_mem: PASS (143.92 s)
 (12/18) guest_limit.qemu.qcow2.virtio_scsi.smp2.virtio_net.RHEL.7.devel.ppc64le.powerkvm-libvirt.limit.without_graphics.max_mem_cpu: PASS (234.79 s)
 (13/18) guest_limit.qemu.qcow2.virtio_scsi.smp2.virtio_net.RHEL.7.devel.ppc64le.powerkvm-libvirt.limit.without_graphics.max_disk: PASS (58.76 s)
 (14/18) guest_limit.qemu.qcow2.virtio_scsi.smp2.virtio_net.RHEL.7.devel.ppc64le.powerkvm-libvirt.limit.without_graphics.max_network: PASS (55.87 s)
 (15/18) guest_limit.qemu.qcow2.virtio_scsi.smp2.virtio_net.RHEL.7.devel.ppc64le.powerkvm-libvirt.limit.without_graphics.max_disk_network: PASS (69.98 s)
 (16/18) guest_limit.qemu.qcow2.virtio_scsi.smp2.virtio_net.RHEL.7.devel.ppc64le.powerkvm-libvirt.limit.without_graphics.max_cpu_disk_network: PASS (84.98 s)
 (17/18) guest_limit.qemu.qcow2.virtio_scsi.smp2.virtio_net.RHEL.7.devel.ppc64le.powerkvm-libvirt.limit.without_graphics.max_cpu_mem_disk_network: PASS (270.90 s)
 (18/18) guest_remove.qemu.qcow2.virtio_scsi.smp2.virtio_net.RHEL.7.devel.ppc64le.powerkvm-libvirt.remove_guest.without_disk: PASS (1.15 s)
RESULTS    : PASS 18 | ERROR 0 | FAIL 0 | SKIP 0 | WARN 0 | INTERRUPT 0 | CANCEL 0
JOB TIME   : 1970.75 s

```

**Comment by sathnaga** [comment]
``` (01/18) guest_import.qemu.qcow2.virtio_scsi.smp2.virtio_net.RHEL.7.devel.ppc64le.powerkvm-qemu.unattended_install.import.import.default_install.aio_native: PASS (40.82 s)
 (02/18) guest_limit.qemu.qcow2.virtio_scsi.smp2.virtio_net.RHEL.7.devel.ppc64le.powerkvm-libvirt.limit.with_graphics.max_cpu: PASS (65.58 s)
 (03/18) guest_limit.qemu.qcow2.virtio_scsi.smp2.virtio_net.RHEL.7.devel.ppc64le.powerkvm-libvirt.limit.with_graphics.max_mem: PASS (240.68 s)
 (04/18) guest_limit.qemu.qcow2.virtio_scsi.smp2.virtio_net.RHEL.7.devel.ppc64le.powerkvm-libvirt.limit.with_graphics.max_mem_cpu: PASS (307.99 s)
 (05/18) guest_limit.qemu.qcow2.virtio_scsi.smp2.virtio_net.RHEL.7.devel.ppc64le.powerkvm-libvirt.limit.with_graphics.max_disk: PASS (64.36 s)
 (06/18) guest_limit.qemu.qcow2.virtio_scsi.smp2.virtio_net.RHEL.7.devel.ppc64le.powerkvm-libvirt.limit.with_graphics.max_network: PASS (73.41 s)
 (07/18) guest_limit.qemu.qcow2.virtio_scsi.smp2.virtio_net.RHEL.7.devel.ppc64le.powerkvm-libvirt.limit.with_graphics.max_disk_network: PASS (78.46 s)
 (08/18) guest_limit.qemu.qcow2.virtio_scsi.smp2.virtio_net.RHEL.7.devel.ppc64le.powerkvm-libvirt.limit.with_graphics.max_cpu_disk_network: PASS (89.96 s)
 (09/18) guest_limit.qemu.qcow2.virtio_scsi.smp2.virtio_net.RHEL.7.devel.ppc64le.powerkvm-libvirt.limit.with_graphics.max_cpu_mem_disk_network: PASS (352.54 s)
 (10/18) guest_limit.qemu.qcow2.virtio_scsi.smp2.virtio_net.RHEL.7.devel.ppc64le.powerkvm-libvirt.limit.without_graphics.max_cpu: PASS (73.80 s)
 (11/18) guest_limit.qemu.qcow2.virtio_scsi.smp2.virtio_net.RHEL.7.devel.ppc64le.powerkvm-libvirt.limit.without_graphics.max_mem: PASS (225.42 s)
 (12/18) guest_limit.qemu.qcow2.virtio_scsi.smp2.virtio_net.RHEL.7.devel.ppc64le.powerkvm-libvirt.limit.without_graphics.max_mem_cpu: PASS (304.55 s)
 (13/18) guest_limit.qemu.qcow2.virtio_scsi.smp2.virtio_net.RHEL.7.devel.ppc64le.powerkvm-libvirt.limit.without_graphics.max_disk: PASS (60.29 s)
 (14/18) guest_limit.qemu.qcow2.virtio_scsi.smp2.virtio_net.RHEL.7.devel.ppc64le.powerkvm-libvirt.limit.without_graphics.max_network: PASS (67.66 s)
 (15/18) guest_limit.qemu.qcow2.virtio_scsi.smp2.virtio_net.RHEL.7.devel.ppc64le.powerkvm-libvirt.limit.without_graphics.max_disk_network: PASS (71.50 s)
 (16/18) guest_limit.qemu.qcow2.virtio_scsi.smp2.virtio_net.RHEL.7.devel.ppc64le.powerkvm-libvirt.limit.without_graphics.max_cpu_disk_network: PASS (81.88 s)
 (17/18) guest_limit.qemu.qcow2.virtio_scsi.smp2.virtio_net.RHEL.7.devel.ppc64le.powerkvm-libvirt.limit.without_graphics.max_cpu_mem_disk_network: PASS (343.09 s)
 (18/18) guest_remove.qemu.qcow2.virtio_scsi.smp2.virtio_net.RHEL.7.devel.ppc64le.powerkvm-libvirt.remove_guest.without_disk: PASS (1.16 s)
RESULTS    : PASS 18 | ERROR 0 | FAIL 0 | SKIP 0 | WARN 0 | INTERRUPT 0 | CANCEL 0
JOB TIME   : 2550.71 s

```

File: libvirt/tests/src/limit.py [hunk 1]
```diff
@@ -0,0 +1,187 @@
...
+        network_str = None
+        disk_str = None
+        # Set network devices
+        if max_network:
+            network_str = "ip link|grep ^[1-9]|wc -l"
+            devices = libvirt_xml.vm_xml.VMXMLDevices()
+            for idx in range(num_network):
+                network = Interface(type_name="bridge")
+                network.mac_address = utils_net.generate_mac_address_simple()
+                network.source = {"bridge": "virbr0"}
```
Comment (nasastry) [inline libvirt/tests/src/limit.py]: Why this hardcoding here? Can this be taken from config file?

Comment (sathnaga) [inline libvirt/tests/src/limit.py]: Sure, can be done

**Comment by sathnaga** [comment]
Dependancy PR: https://github.com/avocado-framework/avocado-vt/pull/1072

File: libvirt/tests/src/limit.py [hunk 2]
```diff
@@ -0,0 +1,187 @@
...
+                vmxml.add_device(disk)
+        vmxml.sync()
+        # Start VM
+        logging.debug("VM XML: \n%s", vmxml)
+        try:
+            vm.start()
+        except virt_vm.VMStartError, detail:
+            for msg in failures.items():
+                if msg[0] in detail:
+                    test.skip("%s", msg[1])
```
Comment (balamuruhans) [inline libvirt/tests/src/limit.py]: `test.cancel()` can be used

File: libvirt/tests/src/limit.py [hunk 3]
```diff
@@ -0,0 +1,187 @@
...
+            test.fail("%s" % detail)
+
+        # Check the memory and vcpu inside guest
+        memtotal = vm.get_totalmem_sys()
+        cpucount = vm.get_cpu_count()
+        session = vm.wait_for_login()
+        if network_str:
+            guestnetworks = int(session.cmd_output(network_str))
+            logging.debug("guestnet: %d", guestnetworks)
+            if (guestnetworks - 2) != num_network:
```
Comment (balamuruhans) [inline libvirt/tests/src/limit.py]: do you consider default interface and `lo` interface for `guestnetworks - 2 `? if we dont have default interface do we get false result ?

File: libvirt/tests/src/limit.py [hunk 4]
```diff
@@ -0,0 +1,187 @@
...
+                failed = True
```
Comment (balamuruhans) [inline libvirt/tests/src/limit.py]: Also we do not consider to check whether interface have ip address?

File: libvirt/tests/src/limit.py [hunk 5]
```diff
@@ -0,0 +1,187 @@
...
+                logging.error("mismatch in guest network devices: \n"
+                              "Expected: %d\nActual: %d", num_network,
+                              guestnetworks)
+        if disk_str:
+            guestdisks = int(session.cmd_output(disk_str))
+            logging.debug("guestdisk: %d", guestdisks)
+            if guestdisks != num_disk:
+                failed = True
```
Comment (balamuruhans) [inline libvirt/tests/src/limit.py]: can we touch the attached disk to validate it is usable from inside guest

File: libvirt/tests/src/limit.py [hunk 6]
```diff
@@ -0,0 +1,187 @@
...
+                logging.error("mismatch in guest disk devices: \n"
+                              "Expected: %d\nActual: %s", num_disk, guestdisks)
+        session.close()
+        guestmem = utils_misc.normalize_data_size("%s G" % guestmemory)
+        # TODO:512 MB thershold deviation value, need to normalize
```
Comment (balamuruhans) [inline libvirt/tests/src/limit.py]: %s/thershold/threshold

File: libvirt/tests/src/limit.py [hunk 7]
```diff
@@ -0,0 +1,187 @@
...
+
+        # Check for host memory and cpu levels and validate against
+        # requested limits, allow to max of 10X for CPU and 2X for memory
+        host_memory = int(memory.rounded_memtotal())
+        if maxmem:
+            if usermaxmem:
+                guestmemory = usermaxmem
+            else:
+                # Normalize to GB
+                guestmemory = int(2.5 * host_memory/(1024 * 1024))
```
Comment (balamuruhans) [inline libvirt/tests/src/limit.py]: in comments requested limits is 2X but here we have 2.5X is there any reason for this

File: libvirt/tests/src/limit.py (hunk 3 above)
Comment (sathnaga) [inline libvirt/tests/src/limit.py]: @balamuruhans Yep, `default +lo` In this test we must have default interface, `lo` is a assumption, I guess that would be in all the cases.

File: libvirt/tests/src/limit.py (hunk 4 above)
Comment (sathnaga) [inline libvirt/tests/src/limit.py]: @balamuruhans ip address check is not the intention of this test, as we want to assert just the max interface per guest

File: libvirt/tests/src/limit.py (hunk 5 above)
Comment (sathnaga) [inline libvirt/tests/src/limit.py]: @balamuruhans same here

File: libvirt/tests/src/limit.py [hunk 8]
```diff
@@ -0,0 +1,187 @@
...
+    max_disk = params.get("max_disk", "no") == "yes"
+    num_network = int(params.get("num_network", 16))
+    num_disk = int(params.get("num_disk", 16))
+    drive_format = params.get("drive_format", "scsi")
+    disk_format = params.get("disk_format", "qcow2")
+    failed = False
+    vmxml = libvirt_xml.VMXML.new_from_inactive_dumpxml(vm_name)
+    org_xml = vmxml.copy()
+    # Destroy the vm
+    vm.destroy()
```
Comment (sathnaga) [inline libvirt/tests/src/limit.py]: Sure

File: libvirt/tests/src/limit.py (hunk 2 above)
Comment (sathnaga) [inline libvirt/tests/src/limit.py]: Sure

File: libvirt/tests/src/limit.py (hunk 6 above)
Comment (sathnaga) [inline libvirt/tests/src/limit.py]: Sure

File: libvirt/tests/src/limit.py (hunk 5 above)
Comment (sathnaga) [inline libvirt/tests/src/limit.py]: can be considered as future `todo` though

**Comment by sathnaga** [comment]
@nasastry @balamuruhans addressed your comments :+1:

File: libvirt/tests/src/limit.py [hunk 9]
```diff
@@ -0,0 +1,188 @@
...
+            test.fail("%s" % detail)
+
+        # Check the memory and vcpu inside guest
+        memtotal = vm.get_totalmem_sys()
+        cpucount = vm.get_cpu_count()
+        session = vm.wait_for_login()
+        if network_str:
+            guestnetworks = int(session.cmd_output(network_str))
+            logging.debug("guestnet: %d", guestnetworks)
+            if (guestnetworks - 2) != num_network:
```
Comment (balamuruhans) [inline libvirt/tests/src/limit.py]: as a pre-setup before test starts for max network test we shall check the interfaces existing and then proceed here with this validation.

File: libvirt/tests/src/limit.py (hunk 5 above)
Comment (balamuruhans) [inline libvirt/tests/src/limit.py]: sure. Also asserting the network interface with ip and with ping test can be considered for todo list

File: libvirt/tests/src/limit.py (hunk 9 above)
Comment (sathnaga) [inline libvirt/tests/src/limit.py]: Sure, will add to TODO

**Comment by sathnaga** [comment]
@balamuruhans can we merge this if no further comments, Thanks

File: libvirt/tests/src/limit.py [hunk 10]
```diff
@@ -0,0 +1,189 @@
...
+    netdst = params.get("netdst", "virbr0")
+    failed = False
+    vmxml = libvirt_xml.VMXML.new_from_inactive_dumpxml(vm_name)
+    org_xml = vmxml.copy()
+    # Destroy the vm
+    vm.destroy()
+    try:
+        # Setup swap
+        if swap_setup:
+            if blk_partition:
```
Comment (balamuruhans) [inline libvirt/tests/src/limit.py]: `if not blk_partition:`

File: libvirt/tests/src/limit.py [hunk 11]
```diff
@@ -0,0 +1,189 @@
...
+                guestmemory = int(2.5 * host_memory/(1024 * 1024))
+        else:
+            pass
+        if not guestmemory:
+            # assign default memory
+            guestmemory = default_mem
+
+        # Set the current and max memory params
+        vmxml.current_mem_unit = 'G'
+        vmxml.max_mem_unit = 'G'
```
Comment (balamuruhans) [inline libvirt/tests/src/limit.py]: it's better not to hardcode, lets have `param` for `mem_unit` defaulting to `G`

File: libvirt/tests/src/limit.py [hunk 12]
```diff
@@ -0,0 +1,189 @@
...
+        vmxml.max_mem = int(guestmemory)
+        vmxml.sync()
+
+        # Set vcpu and topology
+        libvirt_xml.VMXML.set_vm_vcpus(vm_name, max_vcpu, current_vcpu,
+                                       vm_sockets, vm_cores, vm_threads)
+        vmxml = libvirt_xml.VMXML.new_from_inactive_dumpxml(vm_name)
+        # Set vnc display as needed
+        if graphic:
+            if not len(vmxml.get_graphics_devices("vnc")):
```
Comment (balamuruhans) [inline libvirt/tests/src/limit.py]: `len()` not required

File: libvirt/tests/src/limit.py [hunk 13]
```diff
@@ -0,0 +1,189 @@
...
+                graphic = vmxml.get_device_class('graphics')()
+                graphic.add_graphic(vm_name, graphic="vnc")
+        else:
+            if len(vmxml.get_graphics_devices("vnc")):
```
Comment (balamuruhans) [inline libvirt/tests/src/limit.py]: same here

File: libvirt/tests/src/limit.py [hunk 14]
```diff
@@ -0,0 +1,189 @@
...
+                graphic = vmxml.get_device_class('graphics')()
```
Comment (balamuruhans) [inline libvirt/tests/src/limit.py]: this line required irrespective to `vnc` is available in vmxml, lets have it outside the condition

File: libvirt/tests/src/limit.py [hunk 15]
```diff
@@ -0,0 +1,189 @@
...
+                graphic.del_graphic(vm_name)
+        vmxml = libvirt_xml.VMXML.new_from_inactive_dumpxml(vm_name)
+        network_str = None
+        disk_str = None
+        # Set network devices
+        if max_network:
+            network_str = "ip link|grep ^[1-9]|wc -l"
+            devices = libvirt_xml.vm_xml.VMXMLDevices()
```
Comment (balamuruhans) [inline libvirt/tests/src/limit.py]: This variable is not used.

File: libvirt/tests/src/limit.py [hunk 16]
```diff
@@ -0,0 +1,189 @@
...
+            for idx in range(num_network):
+                network = Interface(type_name="bridge")
+                network.mac_address = utils_net.generate_mac_address_simple()
+                network.source = {"bridge": netdst}
+                vmxml.add_device(network)
+        # Set disk devices
+        if max_disk:
+            devices = libvirt_xml.vm_xml.VMXMLDevices()
```
Comment (balamuruhans) [inline libvirt/tests/src/limit.py]: same here

File: libvirt/tests/src/limit.py (hunk 10 above)
Comment (sathnaga) [reply libvirt/tests/src/limit.py]: Sure

File: libvirt/tests/src/limit.py (hunk 11 above)
Comment (sathnaga) [reply libvirt/tests/src/limit.py]: Sure

**Comment by sathnaga** [comment]
@balamuruhans done

**Comment by balamuruhans** [comment]
Thanks @sathnaga

**Comment by sathnaga** [comment]
@balamuruhans Tnx :-)

**Pull request merged**